torch/utils/benchmark/utils/sparse_fuzzer.py

```diff
@@ -0,0 +1,114 @@
+        min_elements: Optional[int] = None,
+        max_elements: Optional[int] = None,
+        dim_parameter: Optional[str] = None,
+        sparse_dim: Optional[str] = None,
+        nnz: Optional[str] = None,
+        density: Optional[str] = None,
+        coalesced: Optional[str] = None,
+        dtype=torch.float32,
+        cuda=False
+    ):
+        """
+        Args:
+            name:
+                A string identifier for the generated Tensor.
+            size:
+                A tuple of integers or strings specifying the size of the generated
+                Tensor. String values will replaced with a concrete int during the
+                generation process, while ints are simply passed as literals.
+            min_elements:
+                The minimum number of parameters that this Tensor must have for a
+                set of parameters to be valid. (Otherwise they are resampled.)
+            max_elements:
+                Like `min_elements`, but setting an upper bound.
+            dim_parameter:
+                The length of `size` will be truncated to this value.
+                This allows Tensors of varying dimensions to be generated by the
+                Fuzzer.
+            sparse_dim: 
+                The number of sparse dimensions in a sparse tensor.
+            nnz: 
+                Number of non-zero elements in the sparse tensor. 
+            density: The `nnz` value will be truncated to this value.
+                This allows tensors of varying sparsities to be generated by the Fuzzer.  
+            coalesced: 
+                The sparse tensor format permits uncoalesced sparse tensors,
+                where there may be duplicate coordinates in the indices. 
+            dtype:
+                The PyTorch dtype of the generated Tensor.
+            cuda:
+                Whether to place the Tensor on a GPU.
+        """
+        super().__init__(name=name, size=size, min_elements=min_elements,
+                         max_elements=max_elements, dim_parameter=dim_parameter, dtype=dtype, cuda=cuda)
+        self._nnz = nnz
+        self._density = density
+        self._coalesced = coalesced
+        self._sparse_dim = sparse_dim
+
+    @staticmethod
+    def sparse_tensor_constructor(size, dtype, sparse_dim, nnz, is_coalesced):
+        if isinstance(size, Number):
+            size = [size] * sparse_dim
+        assert all(size[d] > 0 for d in range(sparse_dim)) or nnz == 0, 'invalid arguments'
+        v_size = [nnz] + list(size[sparse_dim:])
+        if dtype.is_floating_point:
+            v = torch.rand(size=v_size, dtype=dtype, device="cpu")
+        else:
+            v = torch.randint(1, 127, size=v_size, dtype=dtype, device="cpu")
+
+        i = torch.rand(sparse_dim, nnz, device="cpu")
+        i.mul_(torch.tensor(size[:sparse_dim]).unsqueeze(1).to(i))
+        i = i.to(torch.long)
+
+        if not is_coalesced: 
+            v = torch.cat([v, torch.randn_like(v)], 0)
```

Comment:
The nnz or the density vs. coalescing and non-coalescing behavior here seems odd.

When is_coalesced is false the number of elements is doubled!

When is_coalesced is true the number of elements is reduced in the coalescing process by an unclear amount.

Reply:
Thanks, @mruberry I was trying to solve this, and I couldn't figure out  a clear and optimal solution. The main problem here is that the coalescing process will reduce the `nnz` to an unclear amount if the `torch.rand`  generates duplicates values in the process. This could be solved by using a method which generates random value without duplicates, however index values could represent and n-dimensional tensor, so to generate indices without duplicates is too expensive with this restriction.  What I was thinking is keep this behavior  because the probability to generate duplicates indices are low for most of the cases.  

In the other hand when `is_coalesced` is false, yes,  the number of elements is doubled but the number of indices represents the same amount of `nnz` for most of the cases. I mean virtually this represents the same  tensor with the same sparsity pattern.   Moreover, most of the sparse operation will use `coalesce()` method and what we want here is to get a sparse tensor with the same `nnz` even if this is coalesced or not. 

Let me know your thoughts.

Reply:
I appreciate your point. OK, as long as we document this I think it's OK.

Reply:
done